개별 채널 페이지에서 메시지 주고받기 API 호출

srngch opened the pull request:
## 작업 내용

- closes #233

## 리뷰어에게

srngch committed:
add[FE]: Add api to get channel info

srngch committed:
add[FE]: Add api to get messages in channel

srngch committed:
add[FE]: Add api to send messages

hhkim0729 commented on frontend/src/api/api.v1.ts:
```diff
@@ -223,3 +231,30 @@ export const postJoinChannel = async (channelForm: ChannelJoinType) => {
+  if (res.status !== 204) {
+    throw new Error(res.statusText);
+  }
+  return res.data;
```
data가 없어서 그냥 res를 리턴하는 게 맞을 것 같슴다

hhkim0729 commented on frontend/src/api/api.v1.ts:
```diff
@@ -223,3 +231,30 @@ export const postJoinChannel = async (channelForm: ChannelJoinType) => {
+  return res.data;
+};
+
+export const postChannelMessages = async ({
+  channelId,
+  message,
+}: {
+  channelId: number;
+  message: string;
+}) => {
```
타입을 지정하지 않고 이렇게 해도 되는 군요.. 괜히 쓸데없는 타입 자꾸 만들고 있었던 듯

hhkim0729 commented on frontend/src/components/organisms/ChannelDetail.tsx:
```diff
@@ -30,46 +41,58 @@ export default function ChannelDetail({
+    return <div>Error</div>;
+  }
+
+  if (!channel) return <div>채널을 선택해주세요</div>;
```
전에 이렇게 if문 분기하는 거 별로라고 하셔서 하나의 return문에서 && 으로 다 처리하기로 했던 것 같은데 다시 이렇게 가는 게 나은 걸까요...

hhkim0729 replied:
그리고 마지막에 채널이 없는 경우는 로딩 중이거나 오류일 것 같은데 저 문구가 뜨는 다른 상황이 있는 걸까요?

srngch replied:
#256 에서 처리하기

hhkim0729 commented on frontend/src/components/organisms/ChannelDetail.tsx:
```diff
@@ -30,46 +41,58 @@ export default function ChannelDetail({
+  const whoamiQuery = useQuery<UserType, AxiosError>({
+    queryKey: ['whoami'],
+    queryFn: getWhoami,
+  });
```
이전에 query, mutation 등은 페이지 단위에서 처리하고 자식 컴포넌트에 props로 필요한 것을 물려주기로 했던 것 같습니다.
페이지에서 이미 whoami를 호출했으니까 여기서 또 호출할 필요 없이 myUserId를 넘겨줘도 될 것 같아요.

hhkim0729 commented on frontend/src/components/molecule/ChannelUserList.tsx:
```diff
@@ -26,6 +26,15 @@ export default function ChannelUserList({
           초대하기
         </Button>
       )}
+      {myUser && (
+        <ChannelUserItem
+          key={myUser?.id}
+          styles={styles}
+          user={myUser}
+          imageSize={imageSize}
+          myUser={myUser}
+        />
+      )}
```
왜 굳이 ChannelDetail의 users에서 나를 빼고 다시 ChannelUserList에서 수동으로 추가하는 걸까요?

srngch replied:
ChannelDetail 컴포넌트에서 users sort 메소드에 넘기는 compare 함수에서 자기 자신을 구분할 수 없어서요..! 그 전 코드에 테스트용 상수가 넣어져있었습니당

hhkim0729 commented on frontend/src/types/channelUserType.ts:
```diff
@@ -1,12 +1,12 @@
-  owner: 0,
-  admin: 1,
-  normal: 2,
+export const RoleType: { [key: string]: string } = {
```
오호 이렇게 키와 값의 타입을 정해줄 수도 있군요

hhkim0729 commented on frontend/src/components/organisms/Channel.tsx:
```diff
@@ -12,67 +16,78 @@ interface FormData {
+  isLog: boolean;
   text: string;
   createdAt: string;
 }
```
types 디렉토리에 새로 정의를 해주는 것이 좋을 것 같습니다.

hhkim0729 commented on frontend/src/components/molecule/MessageList.tsx:
```diff
@@ -1,26 +1,32 @@
+    queryKey: ['whoami'],
+    queryFn: getWhoami,
+  });
+
```
이것도 props로 myUserId를 받아오는 것이 좋을 것 같습니다

hhkim0729 commented on frontend/src/pages/ChannelPage.tsx:
```diff
@@ -1,72 +1,75 @@
 
   return (
-    <>
-      {channel && (
```
channel이 있는 경우에만 렌더링을 하면 하위 컴포넌트에서 channel이 null인 경우에 발생할 수 있는 문제를 처리해주지 않아도 되기 때문에 코드나 로직이 깔끔해질 것 같은데...

이렇게 하면 아예 헤더가 나오지 않아서 삭제하신 걸까요?
그렇다면 title만 출력되지 않도록 하거나 channel 메인 페이지로 navigate 시키거나 하는 건 어떤가유?

srngch replied:
#256 에서 처리하기

hhkim0729 commented on frontend/src/components/organisms/Channel.tsx:
```diff
@@ -12,67 +16,78 @@ interface FormData {
   const [formData, setFormData] = useState<FormData>({ message: '' });
+  const scrollRef = useRef<HTMLDivElement>(null);
+
+  const getChannelMessagesMutation = useMutation<
+    MessageType[],
+    AxiosError,
+    number
+  >(getChannelMessages);
+
+  const postChannelMessagesMutation = useMutation<
+    any,
+    AxiosError,
+    { channelId: number; message: string }
+  >(postChannelMessages);
```
비동기 로직들 최대한 페이지 단위에서 처리..! 그렇게 하면 useEffect들을 쓰지 않아도 되겠어유

hhkim0729 commented on frontend/src/components/organisms/Channel.tsx:
```diff
@@ -12,67 +16,78 @@ interface FormData {
+  }, [getChannelMessagesMutation.data]);
+
+  useEffect(() => {
+    socket.on('message', (data: any) => {
```
any를 쓰지 않는 방법이 없을지 고민...

srngch replied:
소켓 응답으로 들어오는 객체의 타입(MessageType)으로 맞춰주기로 처리~!

hhkim0729 reviewed (changes requested):
고생하셨습니다~!

srngch committed:
fix[FE]: Fix return type related to 204 response

srngch committed:
move[FE]: Move MessageType to separated file

srngch committed:
refactor[FE]: Move async logic to ChannelPage component

hhkim0729 commented on frontend/src/components/organisms/ChannelDetail.tsx:
```diff
@@ -24,12 +21,12 @@ const findMyUser = (
   if (user1.role !== user2.role) {
     return priority.indexOf(user1.role) - priority.indexOf(user2.role);
   }
+  // 같은 userType이면 nickname 순
```
이제 userType이 아니라 role이라는 단어를 써야 했는데 고치는 걸 깜빡했군요

hhkim0729 commented on frontend/src/components/organisms/ChannelDetail.tsx:
```diff
@@ -38,40 +35,25 @@ const compare = (user1: ChannelUserType, user2: ChannelUserType) => {
   useEffect(() => {
-    if (whoamiQuery.status === 'success' && channel.users) {
-      setMyUser(findMyUser(whoamiQuery.data.id, channel.users));
+    if (channel.users) {
```
 if (!channel.users) return;
이 깔끔할 것 같슴다

hhkim0729 commented on frontend/src/pages/ChannelPage.tsx:
```diff
@@ -37,14 +108,15 @@ export default function ChannelPage() {
   if (whoamiQuery.status === 'error' || getChannelQuery.status === 'error') {
-    return <div>Error</div>;
+    alert('에러가 발생했습니다.');
+    redirect('/channel');
```
redirect()는 loader 내부에서 사용하는 것 같고 보통 useNavigate를 쓰는 것 같은데 크게 차이는 없으려나요..

hhkim0729 commented on frontend/src/components/organisms/Channel.tsx:
```diff
@@ -1,88 +1,23 @@
 interface Props {
-  channel: ChannelType | null;
+  messages: MessageType[];
+  postMessage: (message: string) => void;
```
이벤트 핸들러니까  onSubmit이나 onSubmitMessage 등의 이름을 쓰는 것이 좋을 것 같슴다

srngch replied:
구두로 나눈 이야기 정리..!

- 이벤트 객체를 매개변수로 받는다면 -> 이벤트 핸들러. 함수명 `handleSomething`
- 일반적인 동작을 한다면 -> 일반 함수. 함수명 `doSomething`

srngch committed:
refactor[FE]: Update to increase code readability

- Fix word in comment
- Refactor if statement
- Change `redirect` to `useNavigate`(react-router-dom)
